packages/core-flows/src/region/steps/upsert-and-replace-region-payment-providers.ts

```diff
@@ -0,0 +1,284 @@
+    retrievedPaymentProviderIds
+  )
+
+  if (missingProviders.length) {
+    throw new MedusaError(
+      MedusaError.Types.NOT_FOUND,
+      `Payment providers with ids ${missingProviders.join(
+        ", "
+      )} not found or not enabled`
+    )
+  }
+}
+
+async function getCurrentRegionPaymentProvidersLinks(
+  regionIds: string[],
+  { remoteQuery }: { remoteQuery: RemoteQueryFunction }
+): Promise<
+  {
+    [Modules.REGION]: { region_id: string }
+    [Modules.PAYMENT]: { payment_provider_id: string }
+  }[]
+> {
+  const query = remoteQueryObjectFromString({
+    // TODO: use service: LINKS.RegionPaymentProvider, for that, the links.ts must be moved to the utils package, so that for link querying we can use link const instead of plain strings
+    entryPoint: "region_payment_provider",
+    variables: {
+      filters: { region_id: regionIds },
+      take: null,
+    },
+    fields: ["region_id", "payment_provider_id"],
+  })
+
+  const regionProviderLinks = (await remoteQuery(query)) as {
+    region_id: string
+    payment_provider_id: string
+  }[]
+
+  return regionProviderLinks.map((region) => {
+    return {
+      [Modules.REGION]: {
+        region_id: region.region_id,
+      },
+      [Modules.PAYMENT]: {
+        payment_provider_id: region.payment_provider_id,
+      },
+    }
+  })
+}
+
+async function validateAndNormalizeRegionsProvidersData(
+  input: UpsertAndReplaceRegionPaymentProvidersStepInput["input"],
+  {
+    regionService,
+  }: {
+    regionService: IRegionModuleService
+  }
+): Promise<Map<string, string[]>> {
+  const normalizedInput = Array.isArray(input) ? input : [input]
+  const predicate = "regions" in input[0] ? "regions" : "region_selector"
+
+  const regionPaymentProvidersMap = new Map()
+
+  for (const inputData of normalizedInput) {
+    if ("regions" in inputData) {
+      if (predicate === "region_selector") {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "Invalid input, all input must either be regions or region_selector"
+        )
+      }
+
+      inputData.regions.forEach((region) => {
+        regionPaymentProvidersMap.set(region.id, inputData.payment_providers)
+      })
+    } else if ("region_selector" in inputData) {
+      if (predicate === "regions") {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "Invalid input, all input must either be regions or region_selector"
+        )
+      }
+
+      const regions = await regionService.list(inputData.region_selector, {
+        select: ["id"],
+      })
+
+      regions.forEach((region) => {
+        regionPaymentProvidersMap.set(region.id, inputData.payment_providers)
+      })
+    } else {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Invalid input, must provide either regions or region_selector"
+      )
+    }
+  }
+
+  return regionPaymentProvidersMap
+}
+
+export const upsertAndReplaceRegionPaymentProvidersStepId =
```

Comment:
todo: We are not really upserting anything related to providers or region afaict, so let's call this step something else e.g. `setRegionPaymentProvidersStep`
```suggestion
export const setRegionPaymentProvidersStep =
```